### PR TITLE
allow the actions user to login via the jwt token

### PR DESCRIPTION
**Pull request by bohde**
We have some actions that leverage the Gitea API that began receiving 401 errors, with a message that the user was not found. These actions use the `ACTIONS_RUNTIME_TOKEN` env var in the actions job to authenticate with the Gitea API. The format of this env var in actions jobs changed with go-gitea/gitea/pull/28885 to be a JWT (with a corresponding update to `act_runner`) Since it was a JWT, the OAuth parsing logic attempted to parse it as an OAuth token, and would return user not found, instead of falling back to look up the running task and assigning it to the actions user.

This restores that functionality by parsing Actions JWTs first, and then attempting to parse Oauth JWTs. The code to parse potential old `ACTION_RUNTIME_TOKEN` was kept in case someone is running an older version of act_runner that doesn't support the Actions JWT.

### CODE REVIEW

**Comment by bohde** [comment]
Here's a minimal action workflow that previously worked before the linked PR, running on a server with `setting.Service.RequireSignInView == true` :

```
name: Check Server Version
on: [push]
jobs:
  check-server-version:
    runs-on: ubuntu-latest
    steps:
      - name: Output Server Version
        run: |
          curl -H "Authorization: token $ACTIONS_RUNTIME_TOKEN"  "$GITHUB_SERVER_URL/api/v1/version"
```

File: services/auth/oauth2.go [hunk 1]
```diff
@@ -142,6 +179,15 @@ func (o *OAuth2) Verify(req *http.Request, w http.ResponseWriter, store DataStor
 		return nil, nil
 	}
 
+	user, err := parseActionsJWT(req, store)
```
Comment (wxiaoguang) [inline services/auth/oauth2.go]: I believe it should improve `parseToken` but not introduce a new function with duplicate/inconsistent logic.

Comment (bohde) [reply services/auth/oauth2.go]: Is something like https://github.com/go-gitea/gitea/pull/32527/commits/6e876cc9c630219f892edae5bf7a6a364392b863 what you had in mind?

Comment (wxiaoguang) [reply services/auth/oauth2.go]: It looks better, but where is the permission check? repo1's action token could be used in repo2?

Comment (bohde) [reply services/auth/oauth2.go]: This is just Authentication. Authorization happens with the provided `ActionsTaskID` (e.g. https://github.com/go-gitea/gitea/blob/0d5abd9b3e04a09f5d7de720c99e3451723e028e/routers/api/v1/api.go#L192C24-L192C37). The format of Authentication happens the same here as the previous token format.

Comment (wxiaoguang) [reply services/auth/oauth2.go]: Thank you for the clarification.

The last question: is it possible to write some test code (unit test should be enough) for the new logic?

Comment (bohde) [reply services/auth/oauth2.go]: I added some tests in https://github.com/go-gitea/gitea/pull/32527/commits/f5c0b53d55b9c14b12bd0ba7228f656654e22ac9. Are these what you had in mind?

File: services/auth/oauth2_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	}{
+		"Running":   {TaskID: 47, Expected: true},
+		"Missing":   {TaskID: 1, Expected: false},
+		"Cancelled": {TaskID: 46, Expected: false},
+	}
+
+	for name := range cases {
+		c := cases[name]
+		t.Run(name, func(t *testing.T) {
+			actual := CheckTaskID(context.Background(), c.TaskID)
```
Comment (wxiaoguang) [inline services/auth/oauth2_test.go]: Hmm, it seems quite strange ... what `CheckTaskID` does is "checking whether a task is running"

Maybe it should be named as `CheckTaskIsRunning`

**Review by wxiaoguang — APPROVED** [review]
Overall LGTM except the naming nit.

**Review by Zettat123 — APPROVED** [review]

File: services/auth/oauth2_test.go (hunk 2 above)
Comment (bohde) [reply services/auth/oauth2_test.go]: Renamed in https://github.com/go-gitea/gitea/pull/32527/commits/6aafb51f289de2660484165cdfa100528d5d2852

**Pull request merged**